Update metadata documentation part 4 - reference (#1870)
metoppv/improver: doc/source/metadata/Reference.rst

and has the ``standard_name`` attribute set to ``altitude``.

This **should** be present for the site data to allow
the data to be fully exploited.

axis
****

This CF attribute uses a single capitalised character to indicate
how a coordinate variable should be intepreted.
It can take the values ``X``, ``Y``, ``Z`` and ``T``
representing the three spatial directions and time.

Axis values **should** be set for non-scalar coordinate variables,
which for IMPROVER gridded data are ``X`` and ``Y``.
IMPROVER also has ``threshold`` and ``percentile`` 
coordinate variables, but there are no standards for labelling these.

bounds
******

This CF attribute provides a label pointing to a separate
variable defining the bounds of a point on an axis,
most commonly the start and end of a time period.

This **must** be present if an associated 'bounds variable' exists.

calendar
********

For IMPROVER, this CF attribute is set to ``gregorian`` to indicate
that a Gregorian (standard) calendar is used.

cell_methods
************

This CF attribute can be used to describe the application of simple
statistical processing to a variable 
(e.g. a maximum of the temperature over a period of time).

> It is a string comprising of a list of blank-separated words of the form

Reviewer: ```suggestion
It is a string comprised of a list of blank-separated words of the form ``name: method``. The ``name`` can be a dimension of the variable, a scalar coordinate variable, a valid standard name, or the word ``area``. The ``method`` should be selected from a standard list  described in the `CF Metadata Conventions`_ . It is also possible to include additional information in the form of a ``comment``.
```
I think this can be split up into shorter sentences to improve readability. There were also a couple of full stops missing in the original formulation, which I've added in my suggestion. I've also corrected a typo in 'include' and 'comprised' in this paragraph.

``name: method``, where
the ``name`` can be a dimension of the variable, a scalar coordinate variable,
a valid standard name, or the word ``area`` and
the ``method`` should be selected from a standard list 
described in the `CF Metadata Conventions`_ 
It is also possible to included additional information
in the form of a ``comment``

If any method other than ``point`` is specified for a given axis,
then bounds **must** also be provided for that axis.

For example, ``time: maximum`` would indicate the maximum 

> over the period of the time described by the time bounds.

Reviewer: ```suggestion
over the period of time described by the time bounds.
```

Author: Changed.

Cell methods are covered in more detail in the :ref:`stat-section`.

Conventions
***********

This netCDF attribute specifies a space-separated

> (or, if necessary, comma-separated) 

Reviewer: ```suggestion
(or, if necessary, comma-separated) 
```
Maybe a silly question, but why would a comma-separated list be required?

Author: I think this is if the conventions have spaces in their title.

list of metadata conventions that the file conforms to.
Up until CF version 1.6, 

> strictly only the `CF Metadata Conventions`_ was allowed to be declared here,

Reviewer: ```suggestion
strictly only the `CF Metadata Conventions`_ were allowed to be declared here,
```

Author: Always tricky to know whether to pluralise this, but I've gone with your change.

but a change at 1.7 allowed multiple conventions. 

This **must** be set to include the appropriate version of the CF Convention

> and **should** really include any other conventions that are used

Reviewer: ```suggestion
and **should** include any other conventions that are used
```
I don't think that 'really' is necessary here

Author: Fine.

(although, at present, there is no entry set automatically to indicate the

> extensions used to support enhancementssss used by IMPROVER).

Reviewer: ```suggestion
extensions used to support enhancements used by IMPROVER).
```
I think your key got stuck!

blend_time
**********

This is an IMPROVER-specific variable 

> has been added to indicate when the data was processed (blended)

Reviewer: ```suggestion
which has been added to indicate when the data was processed (blended)
```

Author: changed.

> has been added to indicate when the data was processed (blended)

Author: Deleted unnecessary extra text. 
I've alos moved this section to be just before the "bounds" section, as this is supposed to be alphabetical!

> has been added to indicate when the data was processed (blended)

Author: Deleted unnecessary extra text. 
I've also moved this section to be just before the "bounds" section, as this is supposed to be alphabetical!

to generate this forecast, and can be used to indicate how 'fresh'
the data is.

> This has the ``long_name`` attribute set to ``blend_time``.

Reviewer: ```suggestion
This has the ``long_name`` attribute set to ``blend_time``,
```

but otherwise takes the same form as the ``time`` variable.

Ideally, this **should** be present.

coordinates
***********

This CF attribute lists the scalar coordinates,
i.e. those that do not appear as dimensions of the main variable. 

> This **should** be included where scalar varaiables are present.

Reviewer: ```suggestion
This **should** be included where scalar variables are present.
```

Author: Corrected.

For IMPROVER gridded data this would typically include
``blend_time``, ``height`` and ``time`` 
and for spot data
``altitude``, ``blend_time``, ``latitude``, ``longitude``, 

> ``met_office_site_id`` and ``time wmo_id``.

Author: Good point. The definition of auxiliary coordinates is quite subtle, but I've had a go at re-writing this to be more accurate, without re-producing large chunks of the CF Metadata convention. I think all the common examples in IMPROVER are scalar coordinate variables, but I bet we have some auxiliary coordinates somewhere.

.. warning::

    Use of ``forecast_reference_time`` in IMPROVER is deprecated
    as it is at best unhelpful and at worst it is confusing,
    as IMPROVER generates a blend from multiple sources
    with different start times so there is no unique data time.

forecast_period
***************

This CF variable represents the interval between
the ``forecast_reference_time`` and the validity time (``time``)
and has the ``standard_name`` attribute set to ``forecast_period``.

Ideally, this **should no longer** be used for IMPROVER data.

.. warning::

    Use of ``forecast_period`` in IMPROVER is deprecated
    as it is at best unhelpful and at worst it is confusing,
    as IMPROVER generates a blend from multiple sources
    with different start times so there is no unique data time.

grid_mapping
************

This CF attribute provides a label pointing to a separate 
grid mapping variable, which more fully describes the map projection. 

This **must** be present for gridded data,
as **must** the associated grid mapping variable.

height
******

This CF scalar coordinate variable is included in some
cases to fully describe the quantity of interest.

> This **should** be included if there is any ambiguity in interpretation

Reviewer: ```suggestion
This **should** be included if there is any ambiguity in the interpretation
```

> ofquantity of interest if it is omitted.

Reviewer: ```suggestion
of the quantity of interest if it is omitted.
```

Author: Corrected.

(e.g. an inclusion of ``height`` with a value of ``1.5 m``
for the representatiion of screen level.) 

history
*******

> Ideally, the netCDF attribute it should provide a list of the applications

Reviewer: ```suggestion
Ideally, this netCDF attribute should provide a list of the applications
```

Author: Corrected.

that have modified the original data (i.e. an audit trail),
with recommended practice being to add a date/time stamp
(in the form ``YYYY-MM-DDThh:mm:ssZ``) and identify the software package.
However, in practice, this is far from straightforward for IMPROVER
as it processes a range of model runs,
so there is no single, sequential processing chain
from which to generate such an audit trail,
making it impossible to accurately maintain previous history information.

This is **not currently set** in IMPROVER.

institution
***********

This CF attribute specifies where the original data was produced.

This **must** be present and **should** take the name of the institution from
where the data originated if only data from a single model has been processed.
However, it **should** be set to the institution running the post-processing
for multi-model blended data.

latitude
********

This coordinate variable represents one half of the positional
information for gridded data held on a
Latitude-Longitude (strictly, equirectangular) projection.

> This is also used for site positions, which are are only provided

Author: Removed duplicate

> in latitude and longitude.

Reviewer 2: I believe this is a Met Office convention, not an IMPROVER limitation.

Author: I'm not sure that I would say there is a current Met Office convention in a netCDF context anyway (MONET is so out of date) ... or did you mean a more general convention.

I've not made any change here as I'm think the current statement does not really attribute this, so I don't think it's inaccurate, but feel free to suggest a specific change if you think one is required.

Reviewer 2: Ok. We'll leave this as it is.  I was just thinking that spot data could be defined relative to any grid projection and I don't think IMPROVER would object.

It has the ``standard_name`` attribute set to ``latitude``
and ``units`` set to ``degrees``.
The latitude and longitude can be considered as relative the WGS84
or the World Geodetic System 1984 datum,

> although this is not explicit in the metadata.

Reviewer 2: Would it be better to say "unless explicitly stated in the metadata" as it is possible to describe a different geoid for these spatial coords.

Author: Good thinking - added here and for the "longittude" entry.

This **must** be provided for site data and for gridded data
on a Latitude-Longitude projection.
For gridded data, if any statistical processing over the coordinate 

> has been applied,

Reviewer 2: I think it would be more accurate to say `All data **must** contain either this, or a projection_y_coordinate`. (And something similar for longitude)

there **must** also be an associated ``latitude_bnds`` variable
providing the bounds over which ``cell_methods`` are applied,
although this is often included anyway to define the cell boundaries.

> The ``latitude_bnds`` variable has no attributes as it is tied to the 

Author: I think the sentence only actually references gridded data anyway - I've pulled it into a separate paragraph to make it clearer.

I say "if any statistical processing over the coordinate has been applied", so the cell_methods reference is just by way of expansion rather than being prescriptive.

I can't think of an example of spot data have spatial cell_methods.

main coordinate variable.

least_significant_digit
***********************

This is a variable attribute used by netCDF-writing software to
specify the precision that is maintained when 'bit-shaving'
is applied to provide improved file compression.

> The example value of ``3LL`` indicated that a precision of 3 decimal places

Reviewer: ```suggestion
The example value of ``3LL`` indicates that a precision of 3 decimal places
```

Author: Corrected.

is preserved, i.e. values precise to the nearest 0.001.

> As 'bit-having' is zeroing bits

Reviewer: ```suggestion
As 'bit-shaving' is zeroing bits
```

Author: Corrected.

(that are providing an unrequired level precision),
this would actually be implemented as the power of 2 nearest 0.001.

> This is usually included automatically where the precision limited.

Reviewer: ```suggestion
This is usually included automatically where the precision is limited.
```

Author: Corrected.

The driver for the use of 'bit-shaving' is that although it requires
no extension to the software to read the data (the number formats
in the file are not changed), it facilitates more effective 
reduction in file size, when lossless compression is applied.

long_name
*********

This netCDF-specific variable attribute provides
a descriptive name that is not governed by CF.
If a `CF Standard Name`_ exists for the quantity, 

> this should be used and the ``long_name`` is usually be omitted.

Author: Corrected.

A ``standard_name`` or ``long_name`` **must** be present. 

longitude
*********

This coordinate variable represents one half of the positional
information for gridded data held on a
Latitude-Longitude (strictly, equirectangular) projection.

> This is also used for site positions, which are are only provided

Author: Removed duplicate

in latitude and longitude.
It has the ``standard_name`` attribute set to ``longitude``
and ``units`` set to ``degrees``.
The latitude and longitude can be considered as relative the WGS84
or the World Geodetic System 1984 datum,
although this is not explicit in the metadata.

This **must** be provided for site data and for gridded data
on a Latitude-Longitude projection.
For gridded data, if any statistical processing over the coordinate 
has been applied,
there **must** also be an associated ``longitude_bnds`` variable
providing the bounds over which ``cell_methods`` are applied,
although this is often included anyway to define the cell boundaries.
The ``longitude_bnds`` variable has no attributes as it is tied to the 
main coordinate variable.

met_office_site_id
******************

This IMPROVER-specific coordinate variable
is an 8-character string, zero-padded ID number
used by the Met Office to label all sites.
Within the IMPROVER code, the name is user configurable,
such that it can be changed for different institutions / indices.

Although this precise variable is not appropriate for most users

> other than the Met Office, it is **advisable** to implemented

Author: Corrected.

some form of site identification that has unique elements
and is complete. 

mosg\__
*******

This is intended to indicate a MOSG (Met Office standard grid)
namespace.

> It prefixes atributes to show that they are separate from the 

Reviewer: ```suggestion
It prefixes attributes to show that they are separate from the 
```

Author: Corrected typo

   ``model identifier:cycle time in format yyyymmddTHHMMZ:weight``

percentile
**********

This is an IMPROVER-specific coordinate variable that holds
the set of percentile levels for which values of the variable of
interest are generated.
It has a ``long_name`` attribute set to ``percentile``
and a ``units`` attribute set to ``%``

This **must** be present for percentile variables.

positive
********

Indicates the direction in which values of the vertical coordinate increase,
i.e. where the vertical coordinate is pressure,
the ``positive`` attribute is ``down``.

This **should** be present for vertical coordinates.

projection_x_coordinate
***********************

This coordinate variable represents one half of the positional
information for gridded data held on non-Latitude-Longitude projections.
For example, the Met Office uses a Lambert azimuthal equal area (LAEA) 
projection for the IMPROVER UK domain.
It has a ``standard_name`` attribute set to ``projection_x_coordinate``,
and in the case of the LAEA projection,
the ``units`` attribute is set to ``m``. 
This can be considered as relative to ETRS89
or the European Terrestrial Reference System 1989 

> although this is not explicit in the metadata.

Author: Good point. I did consider removing all references to ETRS89, but as this goes beyond just the ellipsoid to describe fixing reference points on the earth, I think it's interesting to keep it in. However, I've moved all the references to ETRS-89 into the note covering this.

In future, we should add WKT CRS information to StaGE and IMPROVER files, which will directly provide a fuller set of information.

This **must** be provided for gridded data
on a non-Latitude-Longitude projection.
For gridded data, if any statistical processing over the coordinate 
has been applied,
there **must** also be an associated ``projection_x_coordinate_bnds`` variable
providing the bounds over which ``cell_methods`` are applied,
although this is often included anyway to define the cell boundaries.
The ``projection_x_coordinate_bnds`` variable has no attributes
as it is tied to the main coordinate variable.

.. note::
    
    European Terrestrial Reference System 1989 is a a datum
    based on WGS84, but fixed on 1-Jan-1989
    to be anchored to the Eurasian continental plate. 
    This is realised through a TRF
    (the European Terrestrial Reference Frame or ETRF).
    ETRS89 is ideal for a Europe-wide consistent mapping and datasets,
    and is an EU INSPIRE directive standard.

>     In practice, it is close enough WGS84 to make no difference

Reviewer: ```suggestion
    In practice, it is close enough to WGS84 to make no difference
```

Author: added word

    for most applications of post-processed meteorological data.

projection_y_coordinate
***********************

This coordinate variable represents one half of the positional
information for gridded data held on non-Latitude-Longitude projections.
For example, the Met Office uses a Lambert azimuthal equal area (LAEA) grid 
for the IMPROVER UK domain.
It has a ``standard_name`` attribute set to ``projection_y_coordinate``,
and in the case of the LAEA projection,
the ``units`` attribute is set to ``m``. 
This can be considered as relative to ETRS89
or the European Terrestrial Reference System 1989 
although this is not explicit in the metadata.

This **must** be provided for gridded data
on a non-Latitude-Longitude projection.
For gridded data, if any statistical processing over the coordinate 
has been applied,
there **must** also be an associated ``projection_y_coordinate_bnds`` variable
providing the bounds over which ``cell_methods`` are applied,
although this is often included anyway to define the cell boundaries.
The ``projection_y_coordinate_bnds`` variable has no attributes
as it is tied to the main coordinate variable.

> realization

Author: Yes. Done.

***********

This CF coordinate variable is used for indexing ensemble members
and has the ``standard_name`` attribute set to ``realization``.
This is not usually seen in the metadata of IMPROVER output files,

> IMPROVER usually generated probabilities of exceedance or percentiles.

Reviewer: ```suggestion
IMPROVER usually generates probabilities of exceedance or percentiles.
```

Author: Corrected.

However, it will be seen in the input file metadata
and may be seen in the output data ``cell_methods``
where processing has been applied over realizations
(e.g. ``realization: mean`` for mean wind direction).

source
******

This CF attribute specifies the method of production of the original data.

This **must** be present and **should** take the value of the original source

> of the data (typically an NWMP model)

Reviewer: ```suggestion
of the data (typically an NWP model)
```

Author: Removed the extra "M"

when no significant post-processing has been applied.

> However, here significant adjustment of the data has occurred

Reviewer: ```suggestion
However, where significant adjustment of the data has occurred
```

Author: Added the required "w"

or a number input sources have been blended,
it **should** be set to ``IMPROVER``.
Often, careful consideration of when it is appropriate to set this
to reference ``IMPROVER`` is required to avoid the metadata being misleading.
It is probably not worth including a version of the IMPROVER software,
unless this can be reliably supplied.

spot_index
**********

> This is an IMPROVER-specific dimension is used as an increasing integer value

Reviewer: ```suggestion
This IMPROVER-specific dimension is used as an increasing integer value
```

Author: Removed the extra words.

index for sites.

ssp\__
******

This is intended to indicate a SPP (statistical post-processing)
namespace.
It prefixes atributes to show that they are separate from the 
`CF Metadata Conventions`_ attributes.

ssp__relative_to_threshold
**************************

This is an IMPROVER-specific varaible attribute

> indicates the nature of the threshold inequality for a probability

Reviewer: ```suggestion
indicating the nature of the threshold inequality for a probability
```

Author: Corrected.

> and takes one of the four values:

Reviewer: ```suggestion
and takes one of the following four values:
```

Author: Added word

* ``greater_than`` 
* ``greater_than_or_equal_to``
* ``less_than`` 
* ``less_than_or_equal_to``

standard_name
*************

> This CF attribute provides a descriptive name,

Reviewer: ```suggestion
This CF attribute provides a descriptive name
```
I don't think that this comma is necessary

from the governed `CF Standard Name`_ list.

> If no `standard_name`` exists for the quantity, 

Reviewer: ```suggestion
If no ``standard_name`` exists for the quantity, 
```

Author: Added required extra character

a ``long_name`` must be used.

A ``standard_name`` or ``long_name`` **must** be present. 

string5 / string8
*****************

These IMPROVER-specific arbitary constants are used to dimension
the character length of the string variable holding
zero padded WMO identifiers and Met Office identifiers, respectively.

threshold
*********

This is an IMPROVER-specific coordinate variable that holds
the set of values of the variable of interest for which the
probability values are generated.
It has a ``long_name`` attribute set to ``threshold``.

> 

Author: You quute right; my mistake. I've updated this to say "It has either a ``standard_name`` or ``long_name`` attribute set to for the of the variable of interest and appropriate ``units``."

> 

Reviewer 2: What about `var_name="threshold"`? I think this should be mentioned here.

> 

Author: @MoseleyS - I think this is what "This is an IMPROVER-specific coordinate variable..." means. I haven't mentioned ``var_name`` for other variables, so adding it in here would make this entry somewhat special (e.g. under the "time" entry, I don't have a statement saying this has a ``var_name="time"``); ``var_name`` is just the iris cube attribute that holds the NetCDF file variable name, so I would consider the phrase "threshold is a variable" to be equivalent to  ``var_name="threshold"``. Am I missing something?

> 

Reviewer 2: IMPROVER contains code that detects threshold coordinates by their `var_name`. As far as I know, it is the only example of the `var_name` being used in IMPROVER, and therefore seems worth mentioning.

> 

Author: @MoseleyS - That's what I'm missing! Updated the text to reflect this. Ta.

This **must** be present for probability variables.

time
****

This CF Variable provides the time at which the parameter value is valid,
and has a ``standard_name`` attribute set to ``time``.
This is an 64-bit integer in ``units`` of ``seconds since 1970-01-01 00:00:00``

This **must** be present.
If any statistical processing over time has been applied
(e.g. accumulation, maxiumum, etc),
there **must** also be ``time_bnds`` variable
providing the time bounds over which ``cell_methods`` are applied.
``time_bnds`` has no attributes as it is tied to the main time variable.

title
*****

This netCDF global attribute provides a succinct description
of what is in the file and should be something that could be used on a plot
to help describe the data. 

This **must** be present, but there is no generally prescribed form
that is must take.

units
*****

> This netCDF variable attribute provides the units of measurement for the quantity.

Reviewer: ```suggestion
This netCDF variable attribute provides the units of measurement for the quantity
```

Author: Removed unwanted full stop.

in a string form recognised by the Unidata's `UDUNITS package`_

> This **must** be present,and for IMPROVER this **must** be SI units,

Reviewer: ```suggestion
This **must** be present and for IMPROVER this **must** be SI units,
```
I don't think that this comma is necessary

Author: Probably note - removed.

with the exception that ``degrees`` (rather than ``radians``)
are used for wind direction. 

> Non-dimensional quantities, such as IMPROVER probabilities

Reviewer: ```suggestion
Non-dimensional quantities, such as IMPROVER probabilities,
```

Author: It's unlike me not to have enough commas - added it in.

have units set to ``1``.

weather_code
************

This IMPROVER variable provides a weather code in the form of an integer value.
It has a ``long_name`` attribute set to ``weather_code``
and a ``units`` attribute set to ``1``.
It also has ``weather_code`` and ``weather_code_meaning`` attributes
which can used to map code values to a short description;
the values use for the Met Office IMPROVER implementation are
shown in the table below.

.. csv-table:: Met Office weather codes
   :header: "Code", "Description"
   :widths: 5, 15
   :file: weather_codes.csv

wmo_id
******

This IMPROVER-specific coordinate variable
is a 5-character string, zero-padded ID number for WMO sites.

> For non-WMO sites it it set to the string ``None``.

Reviewer: ```suggestion
For non-WMO sites it is set to the string ``None``.
```

Author: Corrected.
